photon-ml/src/main/scala/com/linkedin/photon/ml/function/ObjectiveFunction.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.Vector
+import org.apache.spark.broadcast.Broadcast
+
+import com.linkedin.photon.ml.normalization.NormalizationContext
+
+/**
+ * The base objective function class for an optimization problem.
+ */
+abstract class ObjectiveFunction {
+  type Data
+  type Coefficients
+
```

Comment:
Minor issue and just out of curiosity: why not use generic type here, for example ```ObjectiveFunction[Data, Coefficients]```?